File: python/camicia/camicia.py
"""
Simulate a game of Camicia
"""

def simulate_game(player_a, player_b):
    """
    Simulate a game of Camicia with two players having the given cards.
    :param player_a: A list of card values
    :param player_b: A list of card values
    :returns: Dictionary with keys for  status ("loop" or "finished"), cards (# cards played), and tricks (# tricks)
    """
    royalty = {
        'J': 1,
        'Q': 2,
        'K': 3,
        'A': 4,
    }

    game_status = ''
    cards_played = 0
    count_tricks = 0
    pile = []
    previous_states = []
    turn = 0
    has_moved = [0, 0]

    def collect_pile():
        """
        Collect the pile of cards add them to the player 
        who isn't the current turn. Increment number of tricks
        """
        nonlocal pile
        nonlocal count_tricks
        if 1-turn == 0:
            player_a.extend(pile)
        else:
            player_b.extend(pile)
        pile = []
        count_tricks += 1

    def is_finished():
        """Check if the game is finished due to one player having all the cards
        :returns: True if one player has all the cards (empty pile)
        """
        nonlocal game_status
        total_cards = len(player_a) + len(player_b) + len(pile)
        if len(player_a) == total_cards or len(player_b) == total_cards:
            game_status = 'finished'
            return True
        return False
    
    def no_card():
        """
        Handle condition where the current turn player had no cards to give.
        """
        collect_pile()
        is_finished()

    def update_state():
        """
        Update the list of previous game stats with the current state
        detecting if it is game over due to a loop (current state duplicates previous state)
        """
        nonlocal previous_states
        nonlocal game_status
        move = state_hash(player_a, player_b)
        if move in previous_states:
            game_status = 'loop'
        else:
            previous_states.append(move)

    previous_states.append(state_hash(player_a, player_b)) 
    while len(game_status) == 0:
        has_moved[turn] = 1
        card = next_card(turn, player_a, player_b)
        if len(card) == 0:
            no_card()            
        else:
            pile.append(card)
            cards_played += 1

        if card in royalty:
            penalty = royalty[card]
            recipient = turn
            turn  = 1 - turn
            while penalty > 0:
                has_moved[turn] = 1
                card = next_card(turn, player_a, player_b)
                if len(card) == 0:
                    no_card()
                    break
                pile.append(card)
                cards_played += 1
                penalty-= 1

                if card in royalty:
                    penalty = royalty[card]
                    recipient = turn
                    turn  = 1 - turn
        
            if penalty == 0:
                collect_pile()
                turn = 1 - recipient
        turn = 1 - turn
        if has_moved[0] == has_moved[1] and has_moved[0] != 0:
            update_state()
            has_moved[0] = 0
            has_moved[1] = 0
        
        is_finished()

    return {
        'status': game_status,
        'cards': cards_played,
        'tricks': count_tricks,
    }

def next_card(turn, player_a, player_b):
    """
    Get the next card for the current player
    :param turn: 0 for player_a, and 1 for player_b
    :param player_a: list of cards held by player_a
    :param player_b: list of cards held by player_b
    :returns: next card in the sequence or empty string if unavailable. Card will also be removed from playerA/B's set.
    """
    if turn == 0:
        if len(player_a) == 0:
            return ''
        return player_a.pop(0)
    if len(player_b) == 0:
        return ''
    return player_b.pop(0)
    
def state_hash(player_a, player_b):
    """
    Turn the current state of the game into a hash so we can detect duplicate states 
    and mark the game as a loop
    :param player_a: list of cards held by player_a
    :param player_b: list of cards held by player_b
    :returns: a string which is the hash of the game state. The value of number cards don't matter
    only the position
    """
    numbers = ['2', '3', '4', '5', '6', '7', '8', '9', '10']
    return ''.join(['N' if char in numbers else char for char in player_a]) + '|' + ''.join(['N' if char in numbers else char for char in player_b])
    

File: go/camicia/camicia.go
package camicia

import (
	"maps"
	"slices"
	"strings"
)

type Outcome struct {
	finishes bool
	cards    int
	tricks   int
}

var royalty map[string]int = map[string]int{
	"J": 1,
	"Q": 2,
	"K": 3,
	"A": 4,
}

// Simulate a game of camicia with the given cards for players A and B
// @param playerA: slice of string where each string is a card (A, K, Q, J, 10-2)
// @param playerB: slice of string where each string is a card (A, K, Q, J, 10-2)
// @returns: Outcome struct. This says if the play finished (if not finished it looped), and
// how many cards were played and tricks.
func SimulateGame(playerA, playerB []string) Outcome {
	loops := false
	var ret = Outcome{false, 0, 0}
	royalSuits := slices.Collect(maps.Keys(royalty))

	moves := make([]string, 0)
	turn := 0 // will use (1 - turn) to toggle between 0 and 1.
	flag := []int{0, 0}

	_playerA := make([]string, len(playerA))
	copy(_playerA, playerA)
	_playerB := make([]string, len(playerB))
	copy(_playerB, playerB)
	_pile := make([]string, 0)

	moves = append(moves, State2String(_playerA, _playerB))
	for loops == false && ret.finishes == false {
		var card string
		card, _playerA, _playerB = NextCard(turn, _playerA, _playerB)
		flag[turn] = 1
		if len(card) == 0 {
			_playerA, _playerB, _pile = NullCard(1-turn, _playerA, _playerB, _pile, loops, &ret)
		} else {
			_pile = append(_pile, card)
			ret.cards++
		}

		if len(card) > 0 && slices.Contains(royalSuits, card) {
			penalty := royalty[card]
			recipient := turn
			turn = 1 - turn
			for penalty > 0 {
				var next string
				next, _playerA, _playerB = NextCard(turn, _playerA, _playerB)
				flag[turn] = 1
				if len(next) == 0 {
					_playerA, _playerB, _pile = NullCard(recipient, _playerA, _playerB, _pile, loops, &ret)
					break
				} else {
					_pile = append(_pile, next)
					ret.cards++
					penalty--
				}

				if slices.Contains(royalSuits, next) {
					penalty = royalty[next]
					recipient = turn
					turn = 1 - turn
				}
			}

			if penalty == 0 {
				_playerA, _playerB, _pile = CollectPile(recipient, _playerA, _playerB, _pile, &ret)
				turn = 1 - recipient
			}
		}
		turn = 1 - turn
		if flag[0] == flag[1] && flag[0] == 1 {
			loopFound := false
			loopFound, moves = UpdateMoves(_playerA, _playerB, _pile, moves, &ret)
			loops = loops || loopFound
			flag[0] = 0
			flag[1] = 0
		}
		isFinished(_playerA, _playerB, _pile, loops, &ret)
	}
	return ret
}

// Collect the pile.
// @param turn: Whoes turn was it 0 = playerA, 1 = playerB
// @param playerA: first player's cards
// @param playerB: second player's cards
// @param pile: discard pile
// @param status: Update the game status tricks.
// @returns: new version of playerA, playerB, and pile
func CollectPile(turn int, playerA []string, playerB []string, pile []string, status *Outcome) ([]string, []string, []string) {
	if turn == 0 {
		playerA = slices.Concat(playerA, pile)
	} else {
		playerB = slices.Concat(playerB, pile)
	}
	status.tricks++
	pile = pile[:0]
	return playerA, playerB, pile
}

// Return the next card.
// @param turn: Whoes turn was it 0 = playerA, 1 = playerB
// @param playerA: first player's cards
// @param playerB: second player's cards
// @param pile: discard pile
// @returns: card and the new version of playerA, and playerB
func NextCard(turn int, playerA []string, playerB []string) (string, []string, []string) {
	if turn == 0 {
		if len(playerA) == 0 {
			return "", playerA, playerB
		}
		var next string = playerA[0]
		playerA = playerA[1:]
		return next, playerA, playerB
	}
	if len(playerB) == 0 {
		return "", playerA, playerB
	}
	var next string = playerB[0]
	playerB = playerB[1:]
	return next, playerA, playerB
}

// Check if the game completed.
// @param playerA: first player's cards
// @param playerB: second player's cards
// @param pile: discard pile
// @param loops: has the game looped?
// @param status: Update the game status.
// @returns: true if game over will update status.finishes if one player has all the cards.
func isFinished(playerA []string, playerB []string, pile []string, loops bool, status *Outcome) bool {
	if loops || status.finishes { // Don't break an already set status
		return true
	}

	var total int = len(playerA) + len(playerB) + len(pile)
	if len(playerA) == total || len(playerB) == total {
		status.finishes = true
		return true
	}
	return false
}

// What to do when the drawing player had no card.
// @param turn: Whoes turn was it 0 = playerA, 1 = playerB
// @param playerA: first player's cards
// @param playerB: second player's cards
// @param pile: discard pile
// @param loops: has the game looped
// @param status: Update the game status.
// @returns: new version of playerA, playerB, and pile
func NullCard(turn int, playerA []string, playerB []string, pile []string, loops bool, status *Outcome) ([]string, []string, []string) {
	playerA, playerB, pile = CollectPile(turn, playerA, playerB, pile, status)
	isFinished(playerA, playerB, pile, loops, status)
	return playerA, playerB, pile
}

// Make a state string for the current hands. number cards are treated as if they are any number card
// @param playerA: first player's cards
// @param playerB: second player's cards
// @returns: status string for the players cards.
func State2String(playerA []string, playerB []string) string {
	handA := strings.Builder{}
	handB := strings.Builder{}
	royalSuits := slices.Collect(maps.Keys(royalty))

	for _, card := range playerA {
		if slices.Contains(royalSuits, card) {
			handA.WriteString(card)
		} else {
			handA.WriteString("N")
		}
	}

	for _, card := range playerB {
		if slices.Contains(royalSuits, card) {
			handB.WriteString(card)
		} else {
			handB.WriteString("N")
		}
	}

	return handA.String() + "|" + handB.String()
}

// Collect the pile.
// @param playerA: first player's cards
// @param playerB: second player's cards
// @param pile: discard pile
// @param moves: What moves have already been done.
// @param status: Update the game status tricks.
// @returns: new loops flag, and the updated moves
func UpdateMoves(playerA []string, playerB []string, pile []string, moves []string, status *Outcome) (bool, []string) {
	loops := false
	move := State2String(playerA, playerB)
	if slices.Contains(moves, move) {
		loops = true
	}
	moves = append(moves, move)
	return loops, moves
}
